core/domain/exp_domain.py

```diff
@@ -42,14 +42,18 @@
 
 import bs4
 from typing import (
-    Callable, Dict, Final, List, Literal, Mapping, Optional, Sequence, Set,
-    Tuple, TypedDict, Union, cast, overload)
+    TYPE_CHECKING, Callable, Dict, Final, List, Literal, Mapping, Optional,
+    Sequence, Set, Tuple, TypedDict, Union, cast, overload)
 
 from core.domain import html_cleaner  # pylint: disable=invalid-import-from # isort:skip
 from core.domain import html_validation_service  # pylint: disable=invalid-import-from # isort:skip
 from core.domain import interaction_registry  # pylint: disable=invalid-import-from # isort:skip
 from core.platform import models  # pylint: disable=invalid-import-from # isort:skip
 
+if TYPE_CHECKING:
+    from core.domain import rights_domain
+    from core.domain import user_domain
+
```

Comment:
There was issue of cyclic import in exp_domain (the flow was exp_domain → rights_domain → user_services → auth_services → caching_service → exp_domain (again)).

As we only need them for type annotations and do not execute any functions of rights_domain and user_domain, we can avoid executing them at runtime as the TYPE_CHECKING flag is false during runtime.

Reply:
Hmm, I see. Instead of this, try adding the import under the `if MYPY:` condition. For reference have a look at the `change_domain.py` file. You can also add an comment regarding why you are adding the imports here.
Thanks